ports/nrf/supervisor/internal_flash.c

```diff
@@ -0,0 +1,121 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "supervisor/flash.h"
+
+#include <stdint.h>
+#include <string.h>
+
+#include "extmod/vfs.h"
+#include "extmod/vfs_fat.h"
+#include "py/mphal.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "lib/oofatfs/ff.h"
+
+#include "nrf_nvmc.h"
+
+#ifdef BLUETOOTH_SD
+#include "nrf_sdm.h"
+#endif
+
+// defined in linker
+extern uint32_t __fatfs_flash_start_addr[];
+extern uint32_t __fatfs_flash_length[];
+
+#define NO_CACHE        0xffffffff
+#define FL_PAGE_SZ      4096
+
+uint8_t  _flash_cache[FL_PAGE_SZ] __attribute__((aligned(4)));
+uint32_t _flash_page_addr = NO_CACHE;
+
+
+/*------------------------------------------------------------------*/
+/* Internal Flash API
+ *------------------------------------------------------------------*/
+static inline uint32_t lba2addr(uint32_t block) {
+    return ((uint32_t)__fatfs_flash_start_addr) + block * FILESYSTEM_BLOCK_SIZE;
+}
+
+void supervisor_flash_init(void) {
+}
+
+uint32_t supervisor_flash_get_block_size(void) {
+    return FILESYSTEM_BLOCK_SIZE;
+}
+
+uint32_t supervisor_flash_get_block_count(void) {
+    return ((uint32_t) __fatfs_flash_length) / FILESYSTEM_BLOCK_SIZE ;
+}
+
+// TODO support flashing with SD enabled
+void supervisor_flash_flush(void) {
+    if (_flash_page_addr == NO_CACHE) return;
+
+    // Skip if data is the same
+    if (memcmp(_flash_cache, (void *)_flash_page_addr, FL_PAGE_SZ) != 0) {
+//        _is_flashing = true;
+        nrf_nvmc_page_erase(_flash_page_addr);
+        nrf_nvmc_write_words(_flash_page_addr, (uint32_t *)_flash_cache, FL_PAGE_SZ / sizeof(uint32_t));
+    }
+
+    _flash_page_addr = NO_CACHE;
+}
+
+mp_uint_t supervisor_flash_read_blocks(uint8_t *dest, uint32_t block, uint32_t num_blocks) {
+    uint32_t src = lba2addr(block);
+    memcpy(dest, (uint8_t*) src, FILESYSTEM_BLOCK_SIZE*num_blocks);
+    return 0; // success
+}
+
+mp_uint_t supervisor_flash_write_blocks(const uint8_t *src, uint32_t lba, uint32_t num_blocks) {
+    while (num_blocks) {
+        uint32_t const addr      = lba2addr(lba);
+        uint32_t const page_addr = addr & ~(FL_PAGE_SZ - 1);
+
+        uint32_t count = 8 - (lba % 8); // up to page boundary
+        count = MIN(num_blocks, count);
+
+        if (page_addr != _flash_page_addr) {
+            supervisor_flash_flush();
+
+            // writing previous cached data, skip current data until flashing is done
+            // tinyusb stack will invoke write_block() with the same parameters later on
+            //        if ( _is_flashing ) return;
+
```

Comment:
Above is all commented out. Is there a TODO here, or is this code dead, or should be explained?

Reply:
Not sure. I inherited this from the old code. Sorry its not a move. I'm not sure how git knows them.

Reply:
Its from here: https://github.com/adafruit/circuitpython/blob/master/ports/nrf/internal_flash.c#L119